libgnucash/app-utils/fin.scm

```diff
@@ -185,3 +185,171 @@
+;;    7   | $75,917.25 |  $313.08 | $4,071.92 | $71,845.33 
+;;    8   | $71,845.33 |  $296.28 | $4,088.72 | $67,756.61 
+;;    9   | $67,756.61 |  $279.43 | $4,105.57 | $63,651.04 
+;;   10   | $63,651.04 |  $262.49 | $4,122.51 | $59,528.53
+;;
+;; If you calculate each period sequentially (rounding the interest and balance
+;; at each step), you get:
+;;
+;; PERIOD | Open       | Interest | Principal | End
+;;    1   |$100,000.00 |  $412.39 | $3,972.61 | $96,027.39
+;;    2   | $96,027.39 |  $396.01 | $3,988.99 | $92,038.40 
+;;    3   | $92,038.40 |  $379.56 | $4,005.44 | $88,032.96 
+;;    4   | $88,032.96 |  $363.04 | $4,021.96 | $84,011.00 
+;;    5   | $84,011.00 |  $346.45 | $4,038.55 | $79,972.45 
+;;    6   | $79,972.45 |  $329.80 | $4,055.20 | $75,917.25 
+;;    7   | $75,917.25 |  $313.08 | $4,071.92 | $71,845.33 
+;;    8   | $71,845.33 |  $296.28 | $4,088.72 | $67,756.61 
+;;    9   | $67,756.61 |  $279.42 | $4,105.58 | $63,651.03 <- Different 
+;;   10   | $63,651.03 |  $262.49 | $4,122.51 | $59,528.52 <- still $0.01 off
+;;
+;; =================================
+;;
+;; For the following functions the argument names are:
+;;  py: payment frequency (number of payments per year)
+;;  cy: compounding frequency of the nominal rate (per year)
+;;  iy: nominal annual interest rate
+;;  pv: the present value (opening balance)
+;;  pmt: the size of the periodic payment
+;;  n: the payment period we are asking about (the first payment is n=1)
+;;  places: number of decimal places to round the interest amount to 
+;;          at each payment (999 does no rounding)
+;;
+;; Note: only ordinary annuities are supported (payments at the end of
+;;  each period, not at the beginning of each period)
+;;
+;; Unlike the AMORT worksheet on the BA II Plus, these methods will
+;;  handle the smaller payment (bringing the balance to zero, then
+;;  zeroing future payments)
+;;
+;; The present value (pv) must be non-negative.  If not, the balance will be
+;;  treated as 0.
+;; The payment (pmt) can be positive (paying interest, and hopefully
+;;  reducing the balance each payment), or negative (increasing the balance
+;   each payment).
+;; The payment number (n) must be positive for amort_pmt, amort_ppmt, and
+;;  amort_ipmt.  I.e., the first payment is payment 1.
+;; The payment number (n) must be non-negative for amort_balance.  (In this
+;;  case, payment zero is at the _beginning_ of the first period, so
+;;  amort_balance will just be the initial balance.)
+;; If the above conditions on n are violated, the functions return -1 (#f is
+;;  not used, because it causes gnucash to crash).
+;;
+;; A negative interest rate works (if you can find a lender who charges
+;;  negative rates), but negative compounding frequency, or negative payment
+;;  frequency is a bad idea.
+
+;; Calculate the balance remaining after the nth payment
+;; (n must be greater than or equal to zero)
+(define (gnc:amort_pmt py cy iy pv pmt n places)
```

Comment:
I (somehow) lost 15 lines of code that should have been here:
(define (gnc:amort_balance py cy iy pv pmt n places)
  (if (< pv 0) 0
    (if (< n 0) -1
      (if (and (zero? pv) (>= pmt 0)) 0
        (if (zero? n)
          pv
          (let* ((bal-after-int (amort_balanceAfterInterest pv py cy iy places))
                 (bal-after-pmt (amort_balanceAfterPayment bal-after-int  pmt)))
            (gnc:amort_balance py cy iy bal-after-pmt pmt (- n 1) places )))))))

;; Calculate the size of the nth payment
;; (it will just be pmt for all the payments before the final payment,
;; then less,
;; then zero if you keep trying to make payments)
;; (n must be greater than zero)

I'll work on a patch.

Reply:
It would be nicer to use proper idioms when writing guile. Also I think it would be safer to fix C code to handle `#f` to signify error rather than returning `-1`.

```scheme
(define (gnc:amort_balance py cy iy pv pmt n places)
  (cond
   ((< pv 0) 0)
   ((< n 0) -1)
   ((and (zero? pv) (>= pmt 0)) 0)
   ((zero? n) pv)
   (else
    (let* ((bal-after-int (amort_balanceAfterInterest pv py cy iy places))
           (bal-after-pmt (amort_balanceAfterPayment bal-after-int pmt)))
      (gnc:amort_balance py cy iy bal-after-pmt pmt (- n 1) places)))))
```